### ISSUE TITLE
Gain compensation when attenuating an EQ band

### ISSUE THREAD

**Issue by hapst3r** [mixxxdj/mixxx#11409]
### Bug Description

Hej fellows,

I just updated MIXXX and ever since notice a very strange behaviour. Whenever I turn the EQ knobs DOWN, the overall volume raises until it clips. It seems like -- for some reason -- there is some kind of gain compensation occurring that emphasizes those frequencies not filtered out.

Obviously, what I would expect is that when I turn the EQ knob down, the overall volume is reduced to the extent that a specific frequency band's volume is reduced.

At first I thought this was because of ReplayGain, but disabling it doesn't change a thing (it would have suprised me, I'm using that one for while without having such an issue). 

Then I thought there may have been a change in equalizers and their workings overall (which seems to be indicated by the release notes for 2.3.4) and I experimented with different "Equalizer-Plugin"s as mentioned in the Equalizer part of the Settings. And indeed, when I choose "Biquad-Equalizer", the band knobs work (a bit more) as intended. 

However, I believe that the frequency is attenuated to a much lesser degree. When I turn down all three frequency knobs fully, the volume is halved (and I believe it should be MUCH quieter at that point, IIRC from my other DJ'ing sessions). And indeed, when I kill all three bands, I can still hear stuff rather loudly.

The other EQ models and their behaviour:
- Bessel4 LV-Mix isolator: the problem mentioned, turning down a frequency knob increases overall volume to the point of hardcore clipping.
- Bessel8 LV-Mix Isolator: see above
- Biquad-Equalizer with mute ("mit Stummschaltung" in german): see above
- LinkwitzRiley8 Isolator: insane clipping when all frequency knobs are at +-0. The only way to not have it clipping at a ridiculous(!) degree is to fully turn down the MID and LOW bands to -100%.

Furthermore, I have checked whether the problem was related to me using a DJ controller (Traktor Kontrol S2 mk3) by using my laptop's soundcard. The problems remain the same.

I have checked all Settings pages, but I could not find any setting related to the problem I am facing. And as I said, as a Mixxx-User of ~1,5years, I've never had similar issues.

**EDIT**: Rolling back the system to prior to system (and mixxx) update, the only difference I notice is that by default, "Biquad-Equalizer" (without mute) is activated. All other equalizer models cause problems, and the volume "issue" (or misperception) as described above relating to overall volume when all knobs are fully turned down or the bands are killed persist.

**EDIT2**: Rolling back even further, I feel rather certain that apart from the "equalizer model" problem, there is another issue with how loud overall sound is when all frequency knobs are at -100% (it is louder than it should be and than is has been in earlier versions).

If you need more information, feel free to ask, I'll try to provide what's necessary :-)

### Version

2.3.4

### OS

openSUSE Tumbleweed, up to date

**Comment by ronso0** [comment]
Can you confirm that, besides EQs in the mixer, there is no quick effect and no effect in the effect units activated (or rather neither any deck nor the master output is routed through an effect unit)?

**Comment by daschuer** [comment]
Interesting, I cannot reproduce the issue on Ubuntu Focal. I have probably not the right test? 

I did this: 
* Using the current 2.3 HEAD 2.3.4-18-g9db1ac195f
* Reset the EQ preferences to default: 
  * Biquad Full Kill 
  * High shelf EQ 2.484 kHz 
  * Low Shelf EQ 256 Hz 
* Play Deck A
* Turn down all Deck three EQ knobs
* Volume is lowered as epected
* Change EQ To Bessel4 LV-Mixxx
* EQ knobs are reset to unity. gain back to normal. 
* Turn down all Deck three EQ knobs
* Volume is lowered as epected

Result: no issue.

**Comment by daschuer** [comment]
Maybe you can record a screen-cast showing the issue?

**Comment by hapst3r** [comment]
Below you will find two videos to compare. 

The first video contains a version pre-error, where turning the freq-band knobs down actually attenuates the frequency. Using the kill switches actually kills said band, resulting in a (quasi) silence when all bands are killed. My distro openSUSE displays this Mixxx version as `2.3.4-4.2x86_64`. It is installed via the `multimedia:apps` repository.

**Comment by daschuer** [comment]
Where did you found the package 4.5 and 4.4? 

Here a 1.7 package is provided:
https://software.opensuse.org//download.html?project=multimedia%3Aapps&package=mixxx

Does this make a difference?

**Comment by hapst3r** [comment]
Sorry, I was mistaken: I originally installed it from `multimedia:proaudio`. This repo provides 4.4 and 4.5.

Sadly, using the version you mentioned (from `multimedia:apps`) doesn't make a difference. The problem described in OP persists.

**Comment by ronso0** [comment]
> This video has been archived and is not available to watch at this time

Could you please simply attach it here? Idk the size limit for Github attachments, so to minimize the video file you could record only a section of the screen (the mixer with EQ knonbs and VU meter of one deck). I'm using vokoscreen for such things btw.

**Comment by daschuer** [comment]
Are you able to build form source code? This way we can provide a Mixxx version with additional debug output. 
https://github.com/mixxxdj/mixxx/wiki/Compiling-On-Linux

**Comment by hapst3r** [comment]
> > This video has been archived and is not available to watch at this time
> 
> Could you please simply attach it here? Idk the size limit for Github attachments, so to minimize the video file you could record only a section of the screen (the mixer with EQ knonbs and VU meter of one deck). I'm using vokoscreen for such things btw.

https://user-images.githubusercontent.com/55704994/228442175-f09501ca-f7f6-4d31-bebc-41f4e950d2ff.mp4
https://user-images.githubusercontent.com/55704994/228442165-576a0f2a-a685-4238-a5f1-48d4ff7a6317.mp4

**Comment by daschuer** [comment]
I have still not found the source of your package Mixxx in 'multimedia:proaudio''. Can you provide a Link or has that be improved.  
Than we could examine what happens between 4.4 and 4.5.

**Comment by hapst3r** [comment]
> I have still not found the source of your package Mixxx in 'multimedia:proaudio''. Can you provide a Link or has that be improved. Than we could examine what happens between 4.4 and 4.5.

hej, I am talking about this repository/package: https://software.opensuse.org/download.html?project=multimedia%3Aproaudio&package=mixxx.

I don't know when I will find the time to install from source, I hope it will at some point in the following week. Thanks for your patience :)

**Comment by daschuer** [comment]
Thank you. 
This is the log file: 
https://build.opensuse.org/build/multimedia:proaudio/openSUSE_Tumbleweed/x86_64/mixxx/_log

```
  688s] [ 87%] Building CXX object CMakeFiles/mixxx.dir/src/main.cpp.o
[  688s] /usr/bin/c++ -DMIXXX_BUILD_RELEASE -DNDEBUG -DQT_CONCURRENT_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_SCRIPTTOOLS_LIB -DQT_SCRIPT_LIB -DQT_SQL_LIB -DQT_SVG_LIB -DQT_TABLET_SUPPORT -DQT_TESTCASE_BUILDDIR=\"/home/abuild/rpmbuild/BUILD/mixxx-2.3.4/build\" -DQT_TESTLIB_LIB -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -DQT_XML_LIB -DSFC_SUPPORTS_SET_COMPRESSION_LEVEL -D__BATTERY__ -D__BROADCAST__ -D__BULK__ -D__FFMPEG__ -D__HID__ -D__KEYFINDER__ -D__LILV__ -D__LINUX__ -D__MAD__ -D__MODPLUG__ -D__OPUS__ -D__QTKEYCHAIN__ -D__SNDFILE__ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__VINYLCONTROL__ -D__WV__ -Dx86_64 -I/home/abuild/rpmbuild/BUILD/mixxx-2.3.4/src -I/home/abuild/rpmbuild/BUILD/mixxx-2.3.4/build/src -I/usr/include/ffmpeg -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/googletest/googletest/include -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/portaudio -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/rigtorp/SPSCQueue/include -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/qtscript-bytearray -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/libshout-idjc/include -isystem /usr/include/opus -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtConcurrent -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtOpenGL -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtScript -isystem /usr/include/qt5/QtScriptTools -isystem /usr/include/qt5/QtSql -isystem /usr/include/qt5/QtSvg -isystem /usr/include/qt5/QtTest -isystem /usr/include/qt5/QtXml -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -O2 -g -DNDEBUG -std=gnu++17 -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -MD -MT CMakeFiles/mixxx.dir/src/main.cpp.o -MF CMakeFiles/mixxx.dir/src/main.cpp.o.d -o CMakeFiles/mixxx.dir/src/main.cpp.o -c /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/src/main.cpp
```

I cannot spot anything except that it uses concurrent build flags, that are overridden by our flags as intended. 

This is the log form 2.3.4-4.5x86_6  
Interestingly the are only three revisions there is no trace about the good version 2.3.4-4.2x86_64
Do you have a link for this as well?

**Comment by hapst3r** [comment]
> Thank you. This is the log file: https://build.opensuse.org/build/multimedia:proaudio/openSUSE_Tumbleweed/x86_64/mixxx/_log
> 
> ```
>   688s] [ 87%] Building CXX object CMakeFiles/mixxx.dir/src/main.cpp.o
> [  688s] /usr/bin/c++ -DMIXXX_BUILD_RELEASE -DNDEBUG -DQT_CONCURRENT_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_SCRIPTTOOLS_LIB -DQT_SCRIPT_LIB -DQT_SQL_LIB -DQT_SVG_LIB -DQT_TABLET_SUPPORT -DQT_TESTCASE_BUILDDIR=\"/home/abuild/rpmbuild/BUILD/mixxx-2.3.4/build\" -DQT_TESTLIB_LIB -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -DQT_XML_LIB -DSFC_SUPPORTS_SET_COMPRESSION_LEVEL -D__BATTERY__ -D__BROADCAST__ -D__BULK__ -D__FFMPEG__ -D__HID__ -D__KEYFINDER__ -D__LILV__ -D__LINUX__ -D__MAD__ -D__MODPLUG__ -D__OPUS__ -D__QTKEYCHAIN__ -D__SNDFILE__ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__VINYLCONTROL__ -D__WV__ -Dx86_64 -I/home/abuild/rpmbuild/BUILD/mixxx-2.3.4/src -I/home/abuild/rpmbuild/BUILD/mixxx-2.3.4/build/src -I/usr/include/ffmpeg -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/googletest/googletest/include -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/portaudio -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/rigtorp/SPSCQueue/include -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/qtscript-bytearray -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/libshout-idjc/include -isystem /usr/include/opus -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtConcurrent -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtGui -isystem /usr/include/qt5/QtNetwork -isystem /usr/include/qt5/QtOpenGL -isystem /usr/include/qt5/QtWidgets -isystem /usr/include/qt5/QtScript -isystem /usr/include/qt5/QtScriptTools -isystem /usr/include/qt5/QtSql -isystem /usr/include/qt5/QtSvg -isystem /usr/include/qt5/QtTest -isystem /usr/include/qt5/QtXml -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -O2 -g -DNDEBUG -std=gnu++17 -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Woverloaded-virtual -Wfloat-conversion -Werror=return-type -fPIC -MD -MT CMakeFiles/mixxx.dir/src/main.cpp.o -MF CMakeFiles/mixxx.dir/src/main.cpp.o.d -o CMakeFiles/mixxx.dir/src/main.cpp.o -c /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/src/main.cpp
> ```
> 
> I cannot spot anything except that it uses concurrent build flags, that are overridden by our flags as intended.
> 
> This is the log form 2.3.4-4.5x86_6 Interestingly the are only three revisions there is no trace about the good version 2.3.4-4.2x86_64 Do you have a link for this as well?

Not sure, I believe I searched a similar spot as you did (the opensuse build farm) but I was unable to locate anything which would point to a specific version, unfortunately. However, I looked up on yast software which "builder" built the package, which is "lamb03". Checking the portion of the opensuse build farm [responsible for opensuse tumbleweed](https://build.opensuse.org/packages/mixxx/job_history/multimedia:proaudio/openSUSE_Tumbleweed/x86_64) (my distro), there seems to be only recent build with "lamb03" which is [this one](https://build.opensuse.org/package/show/multimedia:proaudio/mixxx?srcmd5=15d4a148017a7c00fc62caaf96a6b0fa). I hope that helps, but I'm not sure.

> Are you able to build form source code? This way we can provide a Mixxx version with additional debug output. https://github.com/mixxxdj/mixxx/wiki/Compiling-On-Linux

I tried my hand at building but remembered why I kinda hated it :D even though I had mixxx already installed and all related dependencies, when running the

**Comment by daschuer** [comment]
You need to install the devel packages of all dependency
I think they are l listed here: 
https://build.opensuse.org/package/view_file/multimedia:proaudio/mixxx/mixxx.spec?expand=1

**Comment by hapst3r** [comment]
In between my last post and now, I've regularly tried to update mixxx using `zypper update mixxx`, which would've updated the package to version -4.7 (and before that, -4.6). I always had to roll back because the error persisted, which means that I was on -4.2 for a long time.

Today I've managed to build the package from source, but the error persists. Now that I've managed to install the build dependencies (using `zypper source-install -d mixxx`), I shall be able to more regularly make tests builds.

In the meantime, I shall stay on -4.2.

**Comment by ronso0** [comment]
If you're able to build from source you can add debug output to the affected EQ effects yourself, e.g. here https://github.com/mixxxdj/mixxx/blob/main/src/effects/backends/builtin/lvmixeqbase.h

Btw. you can keep a stable version installed while testing local builds directly from the build dir, and also specify a certain settings dir. See https://manual.mixxx.org/2.3/en/chapters/appendix/commandline_dev_tools.html

**Comment by daschuer** [comment]
Does the error persist if you build the HEAD of the 2.3 branch or the head of the 2.4 branch?

**Comment by daschuer** [comment]
In the build log I have found this suspicious part: 
```
[   54s] [  0%] Building C object CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o
[   54s] /usr/bin/cc -DT_LINUX  -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -O2 -g -DNDEBUG -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -MD -MT CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o -MF CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o.d -o CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o -c /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c: In function 'fid_parse':
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2356:34: warning: pointer 'rv_86' used after 'realloc' [-Wuse-after-free]
[   54s]  2356 |           curr= (FidFilter*)((rvp-rv) + tmp);
[   54s]       |                              ~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2354:23: note: call to 'realloc' here
[   54s]  2354 |           tmp= (char*)realloc(rv, (rvp-rv) + xtra);
[   54s]       |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2336:35: warning: pointer 'rv_87' used after 'realloc' [-Wuse-after-free]
[   54s]  2336 |  curr= (FidFilter*)(((char*)curr) - rv + tmp); rv= tmp; }
[   54s]       |                     ~~~~~~~~~~~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2335:40: warning: pointer 'rv_87' used after 'realloc' [-Wuse-after-free]
[   54s]  2335 |  rvend= (rvend-rv) * 2 + tmp; rvp= (rvp-rv) + tmp; \
[   54s]       |                                    ~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2335:15: warning: pointer 'rv_87' used after 'realloc' [-Wuse-after-free]
[   54s]  2335 |  rvend= (rvend-rv) * 2 + tmp; rvp= (rvp-rv) + tmp; \
[   54s]       |         ~~~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2336:35: warning: pointer 'rv_88' used after 'realloc' [-Wuse-after-free]
[   54s]  2336 |  curr= (FidFilter*)(((char*)curr) - rv + tmp); rv= tmp; }
[   54s]       |                     ~~~~~~~~~~~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2335:40: warning: pointer 'rv_88' used after 'realloc' [-Wuse-after-free]
[   54s]  2335 |  rvend= (rvend-rv) * 2 + tmp; rvp= (rvp-rv) + tmp; \
[   54s]       |                                    ~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2335:15: warning: pointer 'rv_88' used after 'realloc' [-Wuse-after-free]
[   54s]  2335 |  rvend= (rvend-rv) * 2 + tmp; rvp= (rvp-rv) + tmp; \
[   54s]       |         ~~~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2336:35: warning: pointer 'rv_90' used after 'realloc' [-Wuse-after-free]
[   54s]  2336 |  curr= (FidFilter*)(((char*)curr) - rv + tmp); rv= tmp; }
[   54s]       |                     ~~~~~~~~~~~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2335:40: warning: pointer 'rv_90' used after 'realloc' [-Wuse-after-free]
[   54s]  2335 |  rvend= (rvend-rv) * 2 + tmp; rvp= (rvp-rv) + tmp; \
[   54s]       |                                    ~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2335:15: warning: pointer 'rv_90' used after 'realloc' [-Wuse-after-free]
[   54s]  2335 |  rvend= (rvend-rv) * 2 + tmp; rvp= (rvp-rv) + tmp; \
[   54s]       |         ~~~~~~^~~~
[   54s] /home/abuild/rpmbuild/BUILD/mixxx-2.3.4/lib/fidlib/fidlib.c:2334:30: note: call to 'realloc' here
[   54s]  2334 | #define INCBUF { tmp= (char*)realloc(rv, (rvend-rv) * 2); if (!tmp) error("Out of memory"); \
[   54s]       |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

When I build it here, I see: 
```
ccache /usr/bin/cc -DT_LINUX  -g   -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -MD -MT CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o -MF CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o.d -o CMakeFiles/fidlib.dir/lib/fidlib/fidlib.c.o   -c ../../lib/fidlib/fidlib.c
```

So we face a bunch of additional flags: 
```
-O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -O2 
```

Do you have them in you local builds as well? Do you see the warnings? 
How do you build you local build? Via plain cmake?

**Comment by uklotzde** [comment]
Similar warnings for [Fedora 37 (GCC 12.2.1)](https://koji.rpmfusion.org/kojifiles/packages/mixxx/2.3.4/1.fc37/data/logs/x86_64/build.log) and [Fedora 38 (GCC 13.0.1)](https://koji.rpmfusion.org/kojifiles/packages/mixxx/2.3.4/2.fc38/data/logs/x86_64/build.log).

**Comment by daschuer** [comment]
I can confirm that this is from a new warning `-Wuse-after-free` included in `-Wall` with gcc 12. 
By default we don't have `-Wall` enabled for fidlib.c 
@uklotzde how is it enabled in your case?

**Comment by uklotzde** [comment]
> @uklotzde how is it enabled in your case?

Default Fedora build settings.

I even need to patch the benchmark tests, otherwise the build would fail: https://github.com/rpmfusion/mixxx/blob/master/disable_werror_in_tests.patch

**Comment by daschuer** [comment]
@hapst3r  I have provided a test branch. You can fetch it via 
```
git checkout -b daschuer-gh11409_2 2.3
git pull https://github.com/daschuer/mixxx.git gh11409_2
```  
Please build it locally and provide the build log. 
Thank you.

**Comment by hapst3r** [comment]
> Does the error persist if you build the HEAD of the 2.3 branch or the head of the 2.4 branch?

@daschuer If I build the HEAD of 2.3 (a.k.a changing the branch via `git checkout 2.3` in the mixxx directory as cloned from this repository), then mixxx does not exhibit this problem. Building from 2.4 fails because `qt5qml` needs to be present for the configure phase to succeed, and I was unable to find that package for openSUSE - apart from windows-related stuff which seemed so wrong that I didn't consider doing it.

**Comment by daschuer** [comment]
In the attached build log, the fidlib.c is build with these flags: 
```
-DT_LINUX  -g -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Wfloat-conversion -Werror=return-type
```
This matches my flags here and seems to work. 

Are you still able to build the failing version at home? Does the build use different flags?

**Comment by hapst3r** [comment]
> In the attached build log, the fidlib.c is build with these flags:
> 
> ```
> -DT_LINUX  -g -pipe -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -Wall -Wextra -Wfloat-conversion -Werror=return-type
> ```
> 
> This matches my flags here and seems to work.
> 
> Are you still able to build the failing version at home? Does the build use different flags?

omg, I feel like an idiot: after building, I always executed `mixxx` - which is the distribution's version and not the one in the `build/`-directory. Re-building and testing (see code below), I notice that none of the builds I said were working actually are working.

To make sure I don't make any (more) foolish mistakes when doing test-builds, this is the sequence I followed for each build (without any "cleanup" afterwards or beforehand).
```
asdf
```

**Comment by daschuer** [comment]
This is your build log https://github.com/mixxxdj/mixxx/files/11240798/230418-gh11409_2-build.log for later reference.

Does the unittest run through? 
Call ctest from ~/mixxx/build

**Comment by hapst3r** [comment]
> This is your build log https://github.com/mixxxdj/mixxx/files/11240798/230418-gh11409_2-build.log for later reference.

Sorry

> Does the unittest run through? Call ctest from ~/mixxx/build

- [ ] 2.3 [230416-2.3-build.log](build.log) [230416-2.3-ctest.log](ctest.log)
```
The following tests did not run:
	205 - EngineBufferE2ETest.SoundTouchToggleTest (Disabled)
	206 - EngineBufferE2ETest.RubberbandToggleTest (Disabled)
	207 - EngineBufferE2ETest.KeylockReverseTest (Disabled)
	566 - SoftTakeoverTest.SuperFastNotSame (Disabled)
	573 - SoftTakeoverTest.PrevNearLess_NewFarLess_Late (Disabled)
	574 - SoftTakeoverTest.PrevNearLess_NewFarMore_Late (Disabled)
	581 - SoftTakeoverTest.PrevNearMore_NewFarLess_Late (Disabled)
	582 - SoftTakeoverTest.PrevNearMore_NewFarMore_Late (Disabled)
	585 - SoftTakeoverTest.PrevFarLess_NewFarLess_Soon (Disabled)
	590 - SoftTakeoverTest.PrevFarLess_NewFarMore_Late (Disabled)
	594 - SoftTakeoverTest.PrevFarMore_NewFarMore_Soon (Disabled)
	597 - SoftTakeoverTest.PrevFarMore_NewFarLess_Late (Disabled)

The following tests FAILED:
	161 - DirectoryDAOTest.relocateDirectory (Failed)
Errors while running CTest
Output from these tests are in: /home/hapster/mixxx/build/Testing/Temporary/LastTest.log
```

**Comment by daschuer** [comment]
Please re-run test using:
```
ctest  --rerun-failed  --output-on-failure
```

**Comment by hapst3r** [comment]
I wonder if it makes sense to simply close this issue, because it seems to be a duplicate and/or subset of #11483 which you already mentioned. Also, I feeli like @uklotzde has a better idea of how to debug things than I do. 

What do you think, is there anything meaningful I can provide that is independent of the other issue?

**Comment by daschuer** [comment]
You test failure is independen to the issue Uwe has reported. Can you give us the output of the line above? 

I am planning to prepare a test build that should give more Infos later. It would be nice if you can try this as well.

**Comment by hapst3r** [comment]
Sure!

Since the test fails in both cases, which HEAD do you want me to test for? Just 2.3 or the test build or both?

**Comment by daschuer** [comment]
Thank you. Just use what you have. If you have both take the 2.3 branch. I don't expect a difference.

**Comment by hapst3r** [comment]
Here you go:
gh11409_2: [230417-gh11409_2-ctest.log](230417-gh11409_2-ctest.log)
2.3: [230417-2.3-ctest.log](230417-2.3-ctest.log)

**Comment by daschuer** [comment]
The posted links are unfortunately dead.

**Comment by hapst3r** [comment]
Sorry, don't know how that has happened...

gh11409_2: [230417-gh11409_2-ctest.log](https://github.com/mixxxdj/mixxx/files/11266202/230417-gh11409_2-ctest.log)
2.3: [230417-2.3-ctest.log](https://github.com/mixxxdj/mixxx/files/11266207/230417-2.3-ctest.log)

**Comment by tymmesyde** [comment]
Experienced the same thing with dnf install on Fedora, first time using this software so i was a little bit confused
Switched to the 2.3.3 flatpak version from gnome software and no more weird EQ issue

**Comment by Swiftb0y** [comment]
Thank you, I think the install from the fedora repos should've been worked around by using Clang on their CI. A proper fix for GCC should've also been merged recently (#11501). Are you able to compile mixxx from source to confirm whether the issue still occurs in the most recent commit on the 2.3 branch?

**Comment by Swiftb0y** [comment]
Thank you for confirming the fix. I'll go ahead and close this issue then.

> For the sake of completeness, I have tried installing mixxx via my package manager (openSUSE Tumbleweed, multimedia:apps repository), and the problem why I have opened this issue is not resolved there.

Does tumbleweed always build the newest 2.3 commit? Afaik other rolling release distros only build releases (and the fix in question has not been part of a release yet) in most cases. I'm not familiar with opensuse's packaging infrastructure so I can't tell you when the release will be broadly accessible (though I assume only after releasing 2.3.5).

**Comment by daschuer** [comment]
@hapst3r 
```
Do you still suffer the test failure? 
/home/hapster/bin/mixxx-testing/src/test/directorydaotest.cpp:238: Failure
Expected equality of these values:
  2
  relocatedTracks.size()
    Which is: 0
```
Should we track it in a new bug report?

**Comment by hapst3r** [comment]
@daschuer yup I do.

`ctest`: [230427-2.3-ctest-lasttest.log](https://github.com/mixxxdj/mixxx/files/11348024/230427-2.3-ctest-lasttest.log)
`ctest --rerun-failed --output-on-failure`: [230427-2.3-ctest-outputonfailure.log](https://github.com/mixxxdj/mixxx/files/11348023/230427-2.3-ctest-outputonfailure.log)
